lib/AST/ASTPrinter.cpp

```diff
@@ -0,0 +1,659 @@
+/*
+ * Copyright (c) 2021-present, Trail of Bits, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed in accordance with the terms specified in
+ * the LICENSE file found in the root directory of this source tree.
+ */
+
+#include "rellic/AST/ASTPrinter.h"
+
+#include <clang/AST/Decl.h>
+#include <clang/AST/Expr.h>
+#include <clang/AST/Stmt.h>
+#include <clang/AST/Type.h>
+#include <gflags/gflags.h>
+#include <glog/logging.h>
+#include <llvm/ADT/SmallVector.h>
+#include <llvm/Support/raw_ostream.h>
+
+#include <sstream>
+
+namespace rellic {
+
+namespace {
+
+static clang::QualType GetBaseType(clang::QualType type) {
+  // FIXME: This should be on the Type class!
+  auto base_type{type};
+  while (!base_type->isSpecifierType()) {
+    if (const auto pty = base_type->getAs<clang::PointerType>())
+      base_type = pty->getPointeeType();
+    else if (const auto bpy = base_type->getAs<clang::BlockPointerType>())
+      base_type = bpy->getPointeeType();
+    else if (const auto aty = clang::dyn_cast<clang::ArrayType>(base_type))
+      base_type = aty->getElementType();
+    else if (const auto fty = base_type->getAs<clang::FunctionType>())
+      base_type = fty->getReturnType();
+    else if (const auto vty = base_type->getAs<clang::VectorType>())
+      base_type = vty->getElementType();
+    else if (const auto pty = base_type->getAs<clang::ParenType>())
+      base_type = pty->desugar();
+    else
+      // This must be a syntax error.
+      break;
+  }
+  return base_type;
+}
+
+static clang::QualType GetDeclType(clang::Decl *decl) {
+  if (auto tdd = clang::dyn_cast<clang::TypedefNameDecl>(decl)) {
+    return tdd->getUnderlyingType();
+  }
+
+  if (auto vd = clang::dyn_cast<clang::ValueDecl>(decl)) {
+    return vd->getType();
+  }
+
+  return clang::QualType();
+}
+
+}  // namespace
+
+void DeclTokenizer::Indent() {
+  for (auto i{0U}; i < indent; ++i) {
+    out.push_back(Token::CreateIndent());
+  }
+}
+
+void DeclTokenizer::PrintPragmas(clang::Decl *decl) {}
+
+void DeclTokenizer::PrintAttributes(clang::Decl *decl) {}
+
+void DeclTokenizer::VisitVarDecl(clang::VarDecl *vdecl) {
+  PrintPragmas(vdecl);
+
+  clang::QualType vtype;
+  if (vdecl->getTypeSourceInfo()) {
+    vtype = vdecl->getTypeSourceInfo()->getType();
+  }
+
+  clang::StorageClass sc{vdecl->getStorageClass()};
+  if (sc != clang::SC_None) {
+    std::string scs_str{clang::VarDecl::getStorageClassSpecifierString(sc)};
+    out.push_back(Token::CreateMisc(scs_str));
+    out.push_back(Token::CreateSpace());
+  }
+
+  switch (vdecl->getTSCSpec()) {
+    case clang::TSCS_unspecified:
+      break;
+
+    case clang::TSCS___thread:
+      out.push_back(Token::CreateMisc("__thread"));
```

Comment:
Can you attach the `vdecl` here and in the misc stuff below? I'd go as far as even including it on the spaces if that can be done.